File: apps/mobile/lib/palette.dart
// The sticky palette, in the one place it becomes a real color.
//
// This mirrors `crates/ymemo-desktop/ui/theme.slint` deliberately, and the two have to keep
// agreeing: the core stores and syncs **the key string alone** (`"yellow"`, `"pink"`, ...),
// so a memo colored on the phone is the same memo colored on the desktop only as long as
// both sides map that key to the same color. Add a key here and there, or not at all.
//
// The colors themselves are the desktop's, minus the translucency: a sticky floats over a
// desktop and reads through it, while a list row and an editor sit on an opaque screen.

import 'package:flutter/material.dart';

/// Palette keys in the order both UIs offer them; the first is the default.
const stickyColors = ['yellow', 'pink', 'green', 'blue', 'purple'];

/// The key a memo or folder carries when it has never been colored, matching
/// `ymemo_core::DEFAULT_COLOR`. An unknown key falls back to it rather than failing, so a
/// vault written by a newer version still draws.
const defaultColor = 'yellow';

/// Body background: the sticky itself, and the editor screen.
Color paletteBg(String key) => switch (key) {
      'pink' => const Color(0xFFFFD6E6),
      'green' => const Color(0xFFD2F0C8),
      'blue' => const Color(0xFFCFE6FF),
      'purple' => const Color(0xFFE6D6F5),
      _ => const Color(0xFFFFF9B1),
    };

/// Title bar: opaque and darker than the body.
Color paletteBar(String key) => switch (key) {
      'pink' => const Color(0xFFF7B8CE),
      'green' => const Color(0xFFB6E0A8),
      'blue' => const Color(0xFFB0D4F2),
      'purple' => const Color(0xFFD0B8EC),
      _ => const Color(0xFFF4E98C),
    };

/// Title bar text, kept legible on [paletteBar] for each color.
Color paletteInk(String key) => switch (key) {
      'pink' => const Color(0xFF7A3350),
      'green' => const Color(0xFF35662F),
      'blue' => const Color(0xFF2A5578),
      'purple' => const Color(0xFF573377),
      _ => const Color(0xFF5C5C25),
    };

/// Saturated accent, for the swatches you pick from and the stripe on a list row.
Color paletteSwatch(String key) => switch (key) {
      'pink' => const Color(0xFFFF9FC0),
      'green' => const Color(0xFF8FD678),
      'blue' => const Color(0xFF7DB8EC),
      'purple' => const Color(0xFFB98FE0),
      _ => const Color(0xFFFFE15C),
    };

/// Background for one row of the memo list.
///
/// [paletteBg] undiluted turns a scrolling list into five bands of poster paint, so the row
/// gets a fraction of it over the surface color — enough to tell two colors apart at a
/// glance, faint enough to read black text on.
Color paletteRow(String key, Color surface) =>
    Color.alphaBlend(paletteBg(key).withValues(alpha: 0.42), surface);

/// The palette as a row of tappable circles, with the current one ringed.
///
/// Both places that change a color use this, so the folder sheet and the editor cannot drift
/// apart in which colors they offer or how a choice looks.
class ColorSwatches extends StatelessWidget {
  const ColorSwatches({super.key, required this.selected, required this.onPick});

  final String selected;
  final void Function(String) onPick;

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisAlignment: MainAxisAlignment.spaceEvenly,
      children: [
        for (final key in stickyColors)
          InkWell(
            onTap: () => onPick(key),
            customBorder: const CircleBorder(),
            child: Padding(
              padding: const EdgeInsets.all(6),
              child: Container(
                width: 34,
                height: 34,
                decoration: BoxDecoration(
                  color: paletteSwatch(key),
                  shape: BoxShape.circle,
                  border: Border.all(
                    color: key == selected ? paletteInk(key) : Colors.black26,
                    width: key == selected ? 3 : 1,
                  ),
                ),
              ),
            ),
          ),
      ],
    );
  }
}

File: apps/mobile/android/app/src/main/kotlin/dev/ymemo/ymemo_mobile/widget/Palette.kt
package dev.ymemo.ymemo_mobile.widget

import android.content.Context
import dev.ymemo.ymemo_mobile.R

/**
 * The sticky palette, for the one place it has to exist outside Dart.
 *
 * This mirrors `lib/palette.dart`, which in turn mirrors the desktop's `theme.slint`. The
 * core syncs **the key alone** (`"yellow"`, `"pink"`, ...), so a memo colored on the desktop
 * is the same color in this widget only as long as every side maps the key the same way. Add
 * a key here and there, or not at all. An unknown key falls back to yellow rather than
 * failing, so a vault written by a newer version still draws.
 *
 * These do not follow the system's dark theme, unlike the widget chrome in `values/colors.xml`:
 * a note is paper, and paper that turned black at sunset would not read as the same note.
 */
internal object Palette {

    /** The paper a note is written on. */
    fun bg(key: String): Int = when (key) {
        "pink" -> 0xFFFFF2F7.toInt()
        "green" -> 0xFFF2FCEF.toInt()
        "blue" -> 0xFFEFF7FF.toInt()
        "purple" -> 0xFFF8F2FD.toInt()
        else -> 0xFFFFFCE3.toInt()
    }

    /** Title bar: opaque and darker than the body. */
    fun bar(key: String): Int = when (key) {
        "pink" -> 0xFFF7B8CE.toInt()
        "green" -> 0xFFB6E0A8.toInt()
        "blue" -> 0xFFB0D4F2.toInt()
        "purple" -> 0xFFD0B8EC.toInt()
        else -> 0xFFF4E98C.toInt()
    }

    /** Title bar text, kept legible on [bar] for each color. */
    fun ink(key: String): Int = when (key) {
        "pink" -> 0xFF7A3350.toInt()
        "green" -> 0xFF35662F.toInt()
        "blue" -> 0xFF2A5578.toInt()
        "purple" -> 0xFF573377.toInt()
        else -> 0xFF5C5C25.toInt()
    }

    /** Saturated accent, for the stripe down the side of a list row. */
    fun swatch(key: String): Int = when (key) {
        "pink" -> 0xFFFF9FC0.toInt()
        "green" -> 0xFF8FD678.toInt()
        "blue" -> 0xFF7DB8EC.toInt()
        "purple" -> 0xFFB98FE0.toInt()
        else -> 0xFFFFE15C.toInt()
    }
}

/**
 * The colors one list widget is drawn in, once its own setting has been taken into account.
 *
 * Left on [WidgetStore.THEME_COLOR] a widget follows the *system's* light/dark setting, the
 * way the widget chrome in `values/colors.xml` always has — a home screen is the launcher's
 * screen, not the app's. Choosing a paper color takes it out of that, and the ink then has to
 * come from the paper: `values-night`'s pale ink on a pastel card would be white on cream.
 */
internal data class Chrome(
    val card: Int,
    val ink: Int,
    val muted: Int,
    val divider: Int,
    val icon: Int,
) {
    companion object {
        fun of(context: Context, key: String): Chrome {
            if (key == WidgetStore.THEME_COLOR) {
                fun color(id: Int) = context.resources.getColor(id, context.theme)
                return Chrome(
                    card = color(R.color.widget_surface),
                    ink = color(R.color.widget_ink),
                    muted = color(R.color.widget_muted),
                    divider = color(R.color.widget_divider),
                    icon = color(R.color.widget_accent),
                )
            }
            val ink = Palette.ink(key)
            return Chrome(
                card = Palette.bg(key),
                ink = ink,
                muted = alpha(ink, 0x9E),
                divider = alpha(ink, 0x2E),
                icon = ink,
            )
        }

        private fun alpha(color: Int, a: Int): Int = (color and 0x00FFFFFF) or (a shl 24)
    }
}
